### PR TITLE
 [tests] Change feature_block.py to use BitcoinTestFramework

### PR DESCRIPTION
**Pull request by jnewbery**
Next step in #10603.

- first three commits tidy up feature_block.py
- fourth commit removes usage of ComparisonTestFramework

Longer term, it would be better to separate net_processing testing from validation testing, but I think this is still a useful PR, since it moves us away from the comparison test framework.

### CODE REVIEW

File: test/functional/invalidblockrequest.py
Comment (JohnVonNeumann) [inline test/functional/invalidblockrequest.py]: Unsure if you guys adhere directly to PEP8 but this comment looks like it could be removed, seems redundant.

File: test/functional/invalidblockrequest.py
Comment (JohnVonNeumann) [inline test/functional/invalidblockrequest.py]: Would imagine this would be best moved inline as opposed to on it's own line to keep convention with inline comments like on line 32 and 29.

File: test/functional/invalidblockrequest.py
Comment (JohnVonNeumann) [inline test/functional/invalidblockrequest.py]: Would assume this block comment should be moved into a docstring block like in test/functional/test_framework/mininode.py P2PStub class. Seems like the standard are great in there, I'm assuming that's down to this section being a refactor.

File: test/functional/invalidtxrequest.py
Comment (JohnVonNeumann) [inline test/functional/invalidtxrequest.py]: Newly added whitespace can be removed.

**Review by JohnVonNeumann — COMMENTED** [review]
Just some basic review of code standards/conventions. Mostly applies to the refactored code, new code looks to have done the job from the get go.

**Comment by jnewbery** [comment]
Thanks for your input @JohnVonNeumann. Three of your nits are on comments that are already in the code, and the last is to revert a pep8 fix (pep8 states that 'inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.')

the invalidtxrequest and invalidblockrequest refactors are covered by #11771 and #11772. If you have functional review of those tests, can you leave comments in those PRs?

**Comment by JohnVonNeumann** [comment]
Is it typically best off to leave preexisting nits for other PRs? Is that something I could fix myself and submit as a PR? Would that be the typical way to address it? I assume you look to keep the PRs clean otherwise people would just keep bringing up little things and it would be difficult to tell what was actually being added in the PR itself?

Apologies on the last PEP8 spacing comment too! I missed that part when reading the docs!

Will have a look at the other PRs and comment in those. Thanks for the feedback as well.

**Comment by jnewbery** [comment]
> Is it typically best off to leave preexisting nits for other PRs? Is that something I could fix myself and submit as a PR? Would that be the typical way to address it? I assume you look to keep the PRs clean otherwise people would just keep bringing up little things and it would be difficult to tell what was actually being added in the PR itself?

I usually clean up style nits as I go, and precede functional-change-commits with tidy-up commits as I have in this PR.

The nits you've raised are more your personal preference than following any particular style convention, so I don't think there's any benefit for you to open a PR for them.

**Comment by JohnVonNeumann** [comment]
Ok this makes sense, although I disagree that all my comments were related directly to my personal preference, for me it was just about keeping the standards throughout the code, for example with the redundant inline comment, I don't see any of them throughout the newly written code, so it would make sense to not have them in the old code. However, you are probably right, and I'll happily defer to your experience. 

I'll try and make my reviews more valuable.

Cheers

**Comment by jnewbery** [comment]
This requires rebase, but I'll leave it as it is until #11771 is reviewed/merged.

File: test/functional/p2p-fullblocktest.py
Comment (ryanofsky) [inline test/functional/p2p-fullblocktest.py]: In commit "Tidy up p2pfullblocktest"

Can you note that this option was removed in the commit message (and maybe say why it wasn't useful?)

**Review by ryanofsky — COMMENTED** [review]
Skipped first two commits which are from #11771 and not up to date. Reviewed remaining individual commits:

- utACK 13ed1ebe08d100500a8f5311c0f91fe4d8f680db [tests] Improve assert message when wait_until() fails
- utACK 1e5da2e20d655d602e62ed87a80df0f23bbe7a34 [tests] Change p2p-fullblocktest to use BitcoinTestFramework
- utACK 63782cd7c7e864f73b2fb715e24513d3d3117214 [tests] Add logging to p2p-fullblocktest.py
- utACK a9554710e65484d7973b706829be67faff53c357 [tests] Tidy up p2pfullblocktest
- utACK 40f7f6b99698552e71537c97e76c2011ebaa5626 [tests] Fix flake8 warnings in p2p-fullblocktest

File: test/functional/p2p-fullblocktest.py
Comment (jnewbery) [reply test/functional/p2p-fullblocktest.py]: Yep. Added to commit message.

**Comment by jnewbery** [comment]
Rebased now that #11771 has been merged.

**Comment by conscott** [comment]
ACK

**Comment by practicalswift** [comment]
utACK 061f4d8cfba58d299dd913bd96f88bddebae8276

Thanks for a very nice cleanup! Looking forward to seeing this merged!

**Comment by jnewbery** [comment]
I saw Travis fail on this, but the trace wasn't printed because of https://github.com/bitcoin/bitcoin/pull/12438.

I reran the job and it passed. It seems that there may be an intermittent failure in the test. I'll rebase on master to get #12438 and try kicking Travis a few times.

**Comment by jnewbery** [comment]
Rebased

File: test/functional/feature_block.py
Comment (MarcoFalke) [inline test/functional/feature_block.py]: nit: Could mention in a comment that this is only needed because `connect_p2p` assumes the thread is already running.

nit: Could rename `connect_p2p` to `reconnect_p2p`, since it only works when the thread is already running?

File: test/functional/feature_block.py
Comment (MarcoFalke) [inline test/functional/feature_block.py]: Why are you changing this comment?

Block 7 spends the output 2 (c.f. line 165)

File: test/functional/feature_block.py
Comment (MarcoFalke) [inline test/functional/feature_block.py]: The comment above `(b12 added last)` is no longer true with this change.

File: test/functional/feature_block.py
Comment (MarcoFalke) [inline test/functional/feature_block.py]: Why are you changing this to `1`? We are building blocks on top of block with id `88`, so naming them `"alt89"` ... seems appropriate.

**Review by maflcko — APPROVED** [review]
Code looks good. Just some comments need update in commit 06d512f11998d9b935edf8b032eb76e491cfb7cc it seems

File: test/functional/feature_block.py
Comment (MarcoFalke) [inline test/functional/feature_block.py]: In commit 06d512f11998d9b935edf8b032eb76e491cfb7cc:

Why are you adding this comment? The `ser_uint256` symbol is not imported and it will unsurprisingly print all zeros, since it was set to 0.

**Comment by MarcoFalke** [comment]
utACK 1f7a8662cdf2c7775942e3abc90c3745c2adaecc

GitHub hiding my review is probably a bug on their end.

File: test/functional/test_framework/util.py
Comment (MarcoFalke) [inline test/functional/test_framework/util.py]: In the last commit:

I think you meant `inspect.getsource`, since the line itself won't help when the file name is unknown. And the Traceback already shows the line number and file name (as well as the predicate, in case it fits into a single line).

File: test/functional/feature_block.py
Comment (jnewbery) [reply test/functional/feature_block.py]: Sounds good. Done both.

File: test/functional/test_framework/util.py
Comment (MarcoFalke) [reply test/functional/test_framework/util.py]: Imo we could do without `inspect`, since it doesn't seem to add information unless I am missing something.

File: test/functional/feature_block.py
Comment (jnewbery) [reply test/functional/feature_block.py]: Yes - you're right. Fixed

File: test/functional/feature_block.py
Comment (jnewbery) [reply test/functional/feature_block.py]: yes - removed

File: test/functional/feature_block.py
Comment (jnewbery) [reply test/functional/feature_block.py]: removed

File: test/functional/test_framework/util.py
Comment (jnewbery) [reply test/functional/test_framework/util.py]: I think this provides useful information when the `wait_until()` fails.

Without this change:

```
2018-03-19T18:29:56.324000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/ubuntu/bitcoin/test/functional/test_framework/test_framework.py", line 126, in main
    self.run_test()
  File "./feature_block.py", line 502, in run_test
    self.sync_blocks([b40], False, 16, b'bad-blk-sigops', reconnect=True)
  File "./feature_block.py", line 1315, in sync_blocks
    self.nodes[0].p2p.send_blocks_and_test(blocks, self.nodes[0], success=success, reject_code=reject_code, reject_reason=reject_reason, request_block=request_block, timeout=timeout)
  File "/home/ubuntu/bitcoin/test/functional/test_framework/mininode.py", line 553, in send_blocks_and_test
    wait_until(lambda: self.reject_code_received == reject_code, lock=mininode_lock, timeout=1)
  File "/home/ubuntu/bitcoin/test/functional/test_framework/util.py", line 222, in wait_until
    assert_greater_than(timeout, time.time())
  File "/home/ubuntu/bitcoin/test/functional/test_framework/util.py", line 42, in assert_greater_than
    raise AssertionError("%s <= %s" % (str(thing1), str(thing2)))
AssertionError: 1521484196.279766 <= 1521484196.3246927
```

With this change:

```
2018-03-19T18:27:45.353000Z TestFramework (INFO): Reject a block with too many P2SH sigops
2018-03-19T18:27:48.557000Z TestFramework.utils (ERROR): wait_until() failed. Predicate: (['            wait_until(lambda: self.reject_code_received == reject_code, lock=mininode_lock, timeout=1)\n'], 553)
2018-03-19T18:27:48.557000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/ubuntu/bitcoin/test/functional/test_framework/test_framework.py", line 126, in main
    self.run_test()
  File "./feature_block.py", line 502, in run_test
    self.sync_blocks([b40], False, 16, b'bad-blk-sigops', reconnect=True)
  File "./feature_block.py", line 1315, in sync_blocks
    self.nodes[0].p2p.send_blocks_and_test(blocks, self.nodes[0], success=success, reject_code=reject_code, reject_reason=reject_reason, request_block=request_block, timeout=timeout)
  File "/home/ubuntu/bitcoin/test/functional/test_framework/mininode.py", line 553, in send_blocks_and_test
    wait_until(lambda: self.reject_code_received == reject_code, lock=mininode_lock, timeout=1)
  File "/home/ubuntu/bitcoin/test/functional/test_framework/util.py", line 227, in wait_until
    raise AssertionError("Predicate {} not true after {} seconds".format(predicate_source, timeout))
AssertionError: Predicate (['            wait_until(lambda: self.reject_code_received == reject_code, lock=mininode_lock, timeout=1)\n'], 553) not true after 1 seconds
```

**Comment by jnewbery** [comment]
Thanks for the review @MarcoFalke - I think I've addressed your comments

**Comment by MarcoFalke** [comment]
Thanks re-utACK 265d7c44b1aae06aee93f745a865807732218a73

**Comment by conscott** [comment]
Test re-ACK 265d7c44b1aae06aee93f745a865807732218a73

**Comment by jamesob** [comment]
Ready for merge?

**Comment by MarcoFalke** [comment]
Needs more (re) acks

**Comment by jnewbery** [comment]
@ryanofsky @practicalswift - care to reACK?

**Review by ryanofsky — COMMENTED** [review]
utACK 265d7c44b1aae06aee93f745a865807732218a73. No code changes since last review except rebase and tweaks in feature_block.py commit (updating comments and reverting b89 loop change). There was also a commit message change mentioning runbarelyexpensive.

File: test/functional/feature_block.py
Comment (jamesob) [inline test/functional/feature_block.py]: I'm pretty sure this functionality gets repeated in several other tests. If so, we should move this somewhere more general and consolidate usages later on.

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/11773/commits/265d7c44b1aae06aee93f745a865807732218a73

File: test/functional/feature_block.py
Comment (jnewbery) [reply test/functional/feature_block.py]: Sure, but let's save that for a future PR.

**Comment by jnewbery** [comment]
4 ACKs. Ready for merge @MarcoFalke ?

File: test/functional/feature_block.py
Comment (jamesob) [reply test/functional/feature_block.py]: Yeah, definitely later.

**Comment by practicalswift** [comment]
utACK 265d7c44b1aae06aee93f745a865807732218a73

**Comment by MarcoFalke** [comment]
@laanwj Could you please merge this, since I am using my travel laptop for the next few days?

**Comment by ajtowns** [comment]
It looks like ebf053ac6135941907ecfebccc778da34b585fac is causing feature_block.py to use about 5GB of RSS memory instead of about 1GB, haven't had time to look into quite why yet. It's causing MemoryError exceptions on my 4GB vm :(

**Comment by jnewbery** [comment]
@ajtowns - I think your memory problem is fixed in #12861

File: test/functional/test_framework/util.py
Comment (sdaftuar) [inline test/functional/test_framework/util.py]: This results in a spurious error message when `wait_until` is used and failure is expected, for instance in the `p2p_segwit.py` test where we advertise a txid from a not-NODE_WITNESS peer and ensure that a subsequence announcement from a NODE_WITNESS peer does not result in a duplicate getdata (while the first request is outstanding).

Comment (jnewbery) [reply test/functional/test_framework/util.py]: Fixed here: #13205

**Pull request merged**
